runtime/common/src/precompile/xtokens.rs

```diff
@@ -0,0 +1,963 @@
+#[derive(RuntimeDebug, Eq, PartialEq, TryFromPrimitive, IntoPrimitive)]
+#[repr(u32)]
+pub enum Action {
+	Transfer = "transfer(address,address,uint256,bytes,uint64)",
+	TransferMultiAsset = "transferMultiAsset(address,bytes,bytes,uint64)",
+	TransferWithFee = "transferWithFee(address,address,uint256,uint256,bytes,uint64)",
+	TransferMultiAssetWithFee = "transferMultiAssetWithFee(address,bytes,bytes,bytes,uint64)",
+	TransferMultiCurrencies = "transferMultiCurrencies(address,(address,uint256)[],uint32,bytes,uint64)",
+	TransferMultiAssets = "transferMultiAssets(address,bytes,uint32,bytes,uint64)",
+}
+
+impl<Runtime> Precompile for XtokensPrecompile<Runtime>
+where
+	Runtime: module_evm::Config + orml_xtokens::Config + module_prices::Config,
+	orml_xtokens::Pallet<Runtime>: XcmTransfer<Runtime::AccountId, Balance, CurrencyId>,
+{
+	fn execute(input: &[u8], target_gas: Option<u64>, _context: &Context, _is_static: bool) -> PrecompileResult {
+		let input = Input::<Action, Runtime::AccountId, Runtime::AddressMapping, Runtime::Erc20InfoMapping>::new(
+			input,
+			target_gas_limit(target_gas),
+		);
+
+		let gas_cost = Pricer::<Runtime>::cost(&input)?;
+
+		if let Some(gas_limit) = target_gas {
+			if gas_limit < gas_cost {
+				return Err(PrecompileFailure::Error {
+					exit_status: ExitError::OutOfGas,
+				});
+			}
+		}
+
+		let action = input.action()?;
+
+		match action {
+			Action::Transfer => {
+				let from = input.account_id_at(1)?;
+				let currency_id = input.currency_id_at(2)?;
+				let amount = input.balance_at(3)?;
+
+				// solidity abi encode bytes will add an offset at input[4]
+				let dest_offset = input.u64_at(4)?;
+				let dest_index = (dest_offset as usize).saturating_div(PER_PARAM_BYTES).saturating_add(1);
+				let dest_bytes_len = input.u32_at(dest_index)?;
+				let mut dest_bytes: &[u8] = &input.bytes_at(dest_index.saturating_add(1), dest_bytes_len as usize)?[..];
+				let versioned_dest: VersionedMultiLocation =
+					Decode::decode(&mut dest_bytes).map_err(|_| PrecompileFailure::Revert {
+						exit_status: ExitRevert::Reverted,
+						output: "invalid dest".into(),
+						cost: target_gas_limit(target_gas).unwrap_or_default(),
+					})?;
+				let dest: MultiLocation = versioned_dest.try_into().map_err(|()| PrecompileFailure::Revert {
+					exit_status: ExitRevert::Reverted,
+					output: "dest bad version".into(),
+					cost: target_gas_limit(target_gas).unwrap_or_default(),
+				})?;
+
+				let weight = input.u64_at(5)?;
+
+				log::debug!(
+					target: "evm",
+					"xtokens: Transfer from: {:?}, currency_id: {:?}, amount: {:?}, dest: {:?}, weight: {:?}",
+					from, currency_id, amount, dest, weight
+				);
+
+				let transferred =
+					<orml_xtokens::Pallet<Runtime> as XcmTransfer<Runtime::AccountId, Balance, CurrencyId>>::transfer(
+						from,
+						currency_id,
+						amount,
+						dest,
+						WeightLimit::Limited(Weight::from_parts(weight, DEFAULT_PROOF_SIZE)),
```

Comment:
I think it is most safe if we just let invoker specify proof size instead of hardcode it here

Reply:
https://github.com/AcalaNetwork/predeploy-contracts/pull/141